### PR TITLE
147 implement switch to room protocol

### PR DESCRIPTION
**Pull request by LucasTesnier**
Rebase of the Hub, Lobby and Room system.

Hub implementation :
- At his creation, the server know handle a hub global loop, which interacts with a communicator instance.
- The hub can handle several clients and dispatch them to different rooms in terms of theirs request.
- Hub can also handle player disconnection / refusal when no places left (or missing performance)

Lobby rebase :
- At his creation a client is now connected to a hub. He receive the list of all the active rooms (Id + name) and can choose to create / join one of this room.
- Client can now create a room and joined it instantly by specifying name of the newly created room. 
- Client can now join a current room by specifying the wanted room id.
- Protocol 2X of communicator is now fully used and the network is switched correctly.

Room rebase :
- At creation, two rooms is available.
- A room need to have a name of strictly 10 characters.
- When a new room is add inside the server, a empty port is now automatically binded.
- Some verification (Like bad ID, same Name, etc...) is implemented to protect the server integrity.
- Everything is now setup to add a graphical interface for replacing the textual one.

WARNING : When this PR's will be merged, multiple client connection and server SigInt will be not functional. This is normal. Everything is gonna return at normal when the rooming process issue will be done.

Enjoy your new triple AAA game ! :)

### CODE REVIEW

File: Server/Server.cpp [hunk 1]
```diff
@@ -14,26 +14,143 @@ using namespace server_data;
...
+    _communicatorInstance.get()->sendDataToAClient(joinDemand.message.clientInfo, nullptr, 0, 11);
+}
+
+void Server::_holdACreateRoomRequest(CommunicatorMessage createDemand)
+{
+    char *tempRoomName = (char *)createDemand.message.data;
+    std::string roomName(11, '\0');
+
+    for (int i = 0; i < 10; i++)
+        roomName[i] = tempRoomName[i];
```
Comment (31Nathan) [inline Server/Server.cpp:116]: strcpy ?

File: Server/Server.hpp [hunk 2]
```diff
@@ -34,22 +37,57 @@ namespace server_data
...
 
         /// @brief Network information of the server. Setup at the construction
         Client _networkInformations;
 
         /// @brief Instance of the communicator library
-        Communicator _communicatorInstance;
+        std::shared_ptr<Communicator> _communicatorInstance;
+
+        /// @brief Cross all machine port and find an empty one
+        /// @param actual port number (+ 100 per iteration)
```
Comment (31Nathan) [inline Server/Server.hpp:65]: In the function it's +101

File: Server/main.cpp [hunk 3]
```diff
@@ -18,6 +18,8 @@ int main(int ac, char **av)
     ArgumentHandler::ServerInformation serverInformation = argumentHandler.extractServerInformation();
     Server server = Server(serverInformation.address, serverInformation.port);
 
-    server.createANewRoom();
+    server.createANewRoom("SleepLucas");
```
Comment (31Nathan) [inline Server/main.cpp:21]: No better Name? xD

**Review by 31Nathan — CHANGES_REQUESTED** [review]

File: Server/Room.hpp [hunk 4]
```diff
@@ -104,6 +116,20 @@ namespace server_data
...
+    /// @return true If the two rooms have the same id
+    /// @return false If the two rooms have different id
+    inline bool operator==(const Room &left, const unsigned short &right) { return ((left.getRoomId() == right)); }
+
+    /// @brief Overload of the == operator to compare rooms with her name
+    /// @param left param of the comparison
+    /// @param right param of the comparison
+    /// @return true If the two rooms have the same name
+    /// @return false If the two rooms have different name
+    inline bool operator==(const Room &left, const std::string &right) { return ((left.getRoomName() == right)); }
```
Comment (ZQUEMA) [inline Server/Room.hpp:132]: Can we compare 2 Rooms ?

File: Server/Server.hpp [hunk 5]
```diff
@@ -34,22 +37,57 @@ namespace server_data
...
+        /// @return the first empty port founded
+        unsigned short _getAFreePort(unsigned short actual);
+
+        /// @brief Start the connexion protocol and ask the server for a place inside the room
+        void _startConnexionProtocol();
+
+        /// @brief Handle a connection request. Send all rooms informations (id + name) to the client.
```
Comment (ZQUEMA) [inline Server/Server.hpp:72]: What is the protocol used here ?

File: Server/main.cpp (hunk 3 above)
Comment (ZQUEMA) [inline Server/main.cpp:21]: Nice reference

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Answer to the questions

File: Server/Server.hpp (hunk 5 above)
Comment (LucasTesnier) [reply Server/Server.hpp:72]: Well done

File: Server/Room.hpp (hunk 4 above)
Comment (LucasTesnier) [reply Server/Room.hpp:132]: :eyes:

File: Server/Server.hpp (hunk 2 above)
Comment (LucasTesnier) [reply Server/Server.hpp:65]: ooops

File: Server/main.cpp (hunk 3 above)
Comment (LucasTesnier) [reply Server/main.cpp:21]: Absolutly not :)

File: Server/Server.cpp (hunk 1 above)
Comment (LucasTesnier) [reply Server/Server.cpp:116]: Not working for char * to std::string

**Review by 31Nathan — APPROVED** [review]

File: Server/Room.hpp (hunk 4 above)
Comment (TomDUVAL-MAHE) [reply Server/Room.hpp:132]: :laughing:

**Review by TomDUVAL-MAHE — APPROVED** [review]

**Review by ZQUEMA — APPROVED** [review]

**Pull request merged**